Make ActiveStorage::Blob keys lowercase

julik opened the pull request:
### Summary

When used with case-insensitive filesystems ActiveStorage blobs might cause non-portable Disk storage directories to be created, and the UNIQUE index that gets configured on the `active_storage_blobs` table is database-dependent in terms of case-sensitivity. If the database index is case-sensitive, but the underlying filesystem is not, blobs might be overwritten and the available key space for uploaded blobs shrinks drastically. This does happen on macOS filesystems already.

This patch changes the ActiveStorage key generation from mixed-case base-58 to lowercase base-36, and increases the key size from 24 bytes to 28. When using base58 there are `58**24` possible key values, and to have the same or more possible keys with base36 we need to have `36**28`, increasing the storage requirement for the key by 4 bytes. Since the database migrations for the blobs table are not CHAR()-size-restricted (and the fix worked for me both on SQLite and MySQL) I believe this is a reasonable compromise.

Note that this will not alter the storage keys that have already been generated, their treatment wills tay unchanged. Neither will it correct the storage directory structures that were created using mixed-case keys on macOS since you need to "rekey" the stored objects if you want the storage directory to be portable to a case-sensitive filesystem.

Closes #34804 , addresses the underlying issue in #33864 for freshly-generated uploaded blobs. Unfortunately it cannot be excluded that for some people some blobs already did get overwritten if SQLite was used in combination with a case-insensitive filesystem and the Disk storage service. Neither does this change existing blob keys.

This change does have an implication that for existing S3 buckets and for other services that potentially use "naive partitioning" (first N  bytes of the given object store key) partition usage is going to become less "fair". However, AWS recently dropped the recommendation on manual key distribution for the key prefix (https://www.infoq.com/news/2018/10/amazon-s3-performance-increase) as it seems they have adopted hashing for the keys the user provides, so this would not be a problem for buckets that get created.

The advantages of using a lowercase-only key:

* Less potential issues with collation-sensitive database indices
* Removes issues with potentially case-insensitive blob stores
* Makes it easier to download blobs from a remote store to a case-insensitive filesystem for debugging

julik committed:
Make ActiveStorage::Blob keys lowercase

julik committed:
Alter test to compensate for key growth

julik committed:
...and off-by-1 errors

georgeclaghorn commented on activestorage/test/models/blob_test.rb:
```diff
@@ -47,6 +47,16 @@ class ActiveStorage::BlobTest < ActiveSupport::TestCase
+    keys.each do |key|
+      assert_match /^[a-z0-9]{28}$/, key
+    end
+  end
```
We can check a single blob’s key and lean on `create_blob`’s defaults:

```ruby
test "create after upload generates a 28-character base36 key" do
  assert_match(/^[a-z0-9]{28}$/, create_blob.key)
end
```

sikachu reviewed (changes requested):
Sorry, but I think `base58.downcase` is not really the right way to fix this.

The reason is that by generating base58 string, we could ended up with `Aaaaa...` and `aaaaa...` which when downcase will cause a key collision.

If we want to make it case insensitive, that means we have to throw away all lowercase or all uppercase characters altogether, which means we need to switch to something else. I was looking around and seems like maybe we can implement [`Base32` then downcase, or `z-base-32`](https://en.wikipedia.org/wiki/Base32) instead.

I'll let @georgeclaghorn decide what we should do next, but `base58.downcase` is definitely not what we're looking for for sure.

julik commented:
@sikachu @georgeclaghorn thank you for your comments. `base58(k).downcase` is a cheap solution indeed, because otherwise I would have to add a base36 implementation to SecureRandom which would increase the API surface there (while not reused anywhere else). I can copy-paste the implementation for `base58` with base36 into the Blob class proper if that will be the preference.

@sikachu as it stands now ActiveStorage::Blob keys are not immune to collisions and this is actually a severe issue, see #34805 - but I do get your point that we should probably use the entropy source better.

julik committed:
Add changelog entry

julik committed:
Simplify test

julik committed:
Use SecureRandom better when generating the key

instead of downcasing the Base58 variant

julik commented:
Why would you want the key to be base64 and/or base32 conformant? It does not give you extra guarantees of uniqueness or the like, but it does force you to do bookkeeping to have N-complement key lengths (28 is 4-complement) and creates those `=`s. Base58  does omit chars that look similar ("lowercase-L versus 1" etc) but I don't know if this has value when you are dealing with a "bag of keys" situation at all.

julik committed:
Double quotes

julik committed:
Update the #key method comment

sikachu commented:
Ah, I agree with you that this is a bag of key situation and hence it does not really matter. I just thought that base32 is closest to what we want to archive (single-case alphabets and numeric) and is one of the standard so adding this to the core would have a high probability of it being reused, despite this is its first usage.

I think I'm going to have to ask @georgeclaghorn if "omitting the similar looking characters" is one of the feature we want or is it just that base58 is already available so we just use it. If it doesn't matter, then maybe base36 will be fine, otherwise I think base32 will be a great addition to the core as well.

julik committed:
Keys keys

keys

georgeclaghorn commented:
Base36 sounds fine to me, but the implementation as it stands now needs to move to [the SecureRandom extension](https://github.com/rails/rails/blob/a796b993ad3717f9a0b3d5be070311ab7a0c95ba/activesupport/lib/active_support/core_ext/securerandom.rb), where it can live alongside `SecureRandom.base58`.

julik committed:
Move base36 into AS core extension

georgeclaghorn commented on activesupport/lib/active_support/core_ext/securerandom.rb:
```diff
@@ -22,4 +24,24 @@ def self.base58(n = 16)
+  #   p SecureRandom.base36 # => "4kugl2pdqmscqtje"
+  #   p SecureRandom.base36(24) # => "77tmhrhjfvfdwodq8w7ev2m7"
+  #
+  def self.base36_lowercase(n = 16)
```
Let’s just call this `base36`. It’s enough to document that it returns a lowercase string (as you already have).

georgeclaghorn commented on activesupport/lib/active_support/core_ext/securerandom.rb:
```diff
@@ -22,4 +24,24 @@ def self.base58(n = 16)
+  #
+  #   p SecureRandom.base36 # => "4kugl2pdqmscqtje"
+  #   p SecureRandom.base36(24) # => "77tmhrhjfvfdwodq8w7ev2m7"
+  #
```
:scissors: this empty line.

georgeclaghorn commented on activesupport/test/core_ext/secure_random_test.rb:
```diff
@@ -19,4 +19,25 @@ def test_base58_with_length
+    assert_match /^[a-z0-9]+$/, s1
+    assert_match /^[a-z0-9]+$/, s2
+  end
+
```
:scissors:

georgeclaghorn commented on activesupport/lib/active_support/core_ext/securerandom.rb:
```diff
@@ -22,4 +24,24 @@ def self.base58(n = 16)
+
+  # SecureRandom.base36 generates a random base36 string in lowercase.
+  #
+  # The argument _n_ specifies the length, of the random string to be generated.
```
I know this came from the `base58` docs, but the comma after “length” is unnecessary. Can you remove it here and above?

georgeclaghorn commented on activesupport/lib/active_support/core_ext/securerandom.rb:
```diff
@@ -22,4 +24,24 @@ def self.base58(n = 16)
+  # If _n_ is not specified or is +nil+, 16 is assumed. It may be larger in the future.
+  # This method can be used over +base58+ if a deterministic case key is necessary.
+  #
+  # The result may contain alphanumeric characters in lowercase
```
The result *will* contain lowercase alphanumeric characters. :wink:

Please add a period to the end of this sentence.

georgeclaghorn commented on activestorage/app/models/active_storage/blob.rb:
```diff
@@ -16,6 +16,8 @@
 # update a blob's metadata on a subsequent pass, but you should not update the key or change the uploaded file.
 # If you need to create a derivative or otherwise change the blob, simply create a new blob and purge the old one.
 class ActiveStorage::Blob < ActiveRecord::Base
+  BASE36_ALPHABET = ("a".."z").to_a + ("0".."9").to_a
```
This constant is no longer used.

julik committed:
Remove unused constant

julik committed:
All the edits

Braces are to avoid warnings when running tests as we start the arguments with a /

julik commented:
🎉 please squash-merge since there are quite a few interim commits

georgeclaghorn commented:
Thank you! :heart:

julik commented on activesupport/lib/active_support/core_ext/securerandom.rb:
```diff
@@ -4,22 +4,42 @@
+  #   p SecureRandom.base36(24) # => "77tmhrhjfvfdwodq8w7ev2m7"
+  def self.base36(n = 16)
+    SecureRandom.random_bytes(n).unpack("C*").map do |byte|
+      idx = byte % 64
```
Might be, I would actually rewrite both base58 and base36 to use a much more pedestrian

```
n.times.map { alphabet[SecureRandom.random_number(alphabet.length) }.join
```

More method calls but probably "less clever" is better in this case. I didn't touch the implementation because I was thinking "this byte trickery for obtaining bytes from a potentially-slow SecureRandom is here for a reason"

julik replied:
Just didn't want to touch it since this is not what the PR is about and this would need microbenchmarks etc.

julik replied:
Noted, I'll open a patch for this once I'm done with #34827
